src/dcegm/egm.py

```diff
@@ -0,0 +1,458 @@
+        savings_grid=exogenous_savings_grid,
+        income_shock=income_shock_draws,
+    )
+    # Interpolate next period policy and values to match the
+    # contemporary matrix of potential next period wealths
+    child_policy = get_child_state_choice_specific_policy(
+        child_node_choice_set,
+        next_period_wealth,
+        next_period_policy=choice_policies_child,
+    )
+    choice_child_values = get_child_state_choice_specific_values(
+        child_node_choice_set,
+        next_period_wealth=next_period_wealth,
+        next_period_value=choice_values_child,
+        compute_value=compute_value,
+    )
+
+    child_state_marginal_utility = get_child_state_policy(
+        child_node_choice_set,
+        next_period_policy=child_policy,
+        next_period_value=choice_child_values,
+        taste_shock_scale=taste_shock_scale,
+        compute_marginal_utility=compute_marginal_utility,
+    )
+
+    child_state_log_sum = calc_exp_max_value(choice_child_values, taste_shock_scale)
+
+    marginal_utility_weighted = (
+        child_state_marginal_utility.reshape(
+            exogenous_savings_grid.shape[0], income_shock_draws.shape[0]
+        )
+        @ income_shock_weights
+    )
+
+    log_sum_weighted = (
+        child_state_log_sum.reshape(
+            exogenous_savings_grid.shape[0], income_shock_draws.shape[0]
+        )
+        @ income_shock_weights
+    )
+
+    return marginal_utility_weighted, log_sum_weighted
+
+
+def calc_exp_max_value(
+    choice_specific_values: np.ndarray, taste_shock_scale: float
+) -> np.ndarray:
+    """Calculate the expected max value given choice specific values.
+
+    With the general extreme value assumption on the taste shocks, this reduces
+    to the log-sum.
+
+    The log-sum formula may also be referred to as the 'smoothed max function',
+    see eq. (50), p. 335 (Appendix).
+
+    Args:
+        choice_specific_values (np.ndarray): Array containing values of the
+            choice-specific value function.
+            Shape (n_choices, n_quad_stochastic * n_grid_wealth).
+        taste_shock_scale (float): Taste shock scale parameter.
+
+    Returns:
+        logsum (np.ndarray): Log-sum formula inside the expected value function.
+            Array of shape (n_quad_stochastic * n_grid_wealth,).
+
+    """
+    col_max = np.amax(choice_specific_values, axis=0)
+    choice_specific_values_scaled = choice_specific_values - col_max
+
+    # Eq. (14), p. 334 IJRS (2017)
+    logsum = col_max + taste_shock_scale * np.log(
+        np.sum(np.exp(choice_specific_values_scaled / taste_shock_scale), axis=0)
+    )
+
+    return logsum
+
+
+def get_child_state_policy(
+    child_node_choice_set: np.ndarray,
+    next_period_policy: np.ndarray,
+    next_period_value: np.ndarray,
+    compute_marginal_utility: Callable,
+    taste_shock_scale: float,
+) -> np.ndarray:
+    """We aggregate the marginal utility of the discrete choices in the next period with
```

Comment:
Given the function name ```get_child_state_policy```, this docstring is not what I would expect. Should adapt one or the other.

Reply:
Renamed the function. It does only calculate the marginal utilities.